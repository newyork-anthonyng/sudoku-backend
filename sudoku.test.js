/**
 * @jest-environment node
 */

const Sudoku = require("./sudoku");
const { rest } = require("msw");
const { setupServer } = require("msw/node");

const BASE_URL = "http://www.cs.utep.edu/cheon/ws/sudoku/new/";

const handlers = [
  rest.get(BASE_URL, (_, res, ctx) => {
    return res(
      ctx.json({
        response: true,
        size: 9,
        squares: [{ x: 0, y: 0, value: 9 }],
      })
    );
  }),
];

const server = setupServer(...handlers);

beforeAll(() => {
  server.listen();
});

afterEach(() => {
  server.resetHandlers();
});

afterAll(() => {
  server.close();
});

it("should create new game", async () => {
  const sudoku = new Sudoku();

  const grid = await sudoku.createNewPuzzle();
  expect(grid).toMatchSnapshot();
});

it("should return null when api call fails", async () => {
  server.use(
    rest.get(BASE_URL, (req, res, ctx) => {
      return res(ctx.status(500));
    })
  );

  const sudoku = new Sudoku();

  const grid = await sudoku.createNewPuzzle();
  expect(grid).toEqual(null);
});

describe("move", () => {
  let sudoku;
  beforeEach(async () => {
    sudoku = new Sudoku();

    await sudoku.createNewPuzzle();
  });
  it("should update grid when player moves", () => {
    const newMove = { x: 0, y: 1, value: 8 };
    sudoku.move(newMove);

    const grid = sudoku.getGrid();
    expect(grid[newMove.x][newMove.y].value).toEqual(newMove.value);
  });

  it("should not update filled grid item", () => {
    const newMove = { x: 0, y: 0, value: 2 };
    sudoku.move(newMove);

    const grid = sudoku.getGrid();
    expect(grid[newMove.x][newMove.y].value).not.toEqual(newMove.value);
  });

  it("should not allow invalid numbers", () => {
    const bigNumber = { x: 0, y: 1, value: 10 };
    sudoku.move(bigNumber);
    let grid = sudoku.getGrid();
    expect(grid[bigNumber.x][bigNumber.y].value).not.toEqual(bigNumber.value);

    const negativeNumber = { x: 0, y: 1, value: -1 };
    sudoku.move(negativeNumber);
    grid = sudoku.getGrid();

    expect(grid[negativeNumber.x][negativeNumber.y].value).not.toEqual(
      negativeNumber.value
    );
  });

  it("should handle when moves are outside grid", () => {
    const originalGrid = sudoku.getGrid();

    const outsideGridMove = { x: 10, y: 10, value: 1 };
    sudoku.move(outsideGridMove);

    const grid = sudoku.getGrid();
    for (let i = 0; i < originalGrid.length; i++) {
      for (let j = 0; j < originalGrid.length; j++) {
        const originalSquare = originalGrid[i][j];
        const square = grid[i][j];
        expect(originalSquare).toEqual(square);
      }
    }
  });
});
