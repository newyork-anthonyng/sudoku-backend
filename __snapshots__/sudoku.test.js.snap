// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create new game 1`] = `
Array [
  Array [
    Object {
      "filled": true,
      "value": 9,
      "x": 0,
      "y": 0,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 0,
      "y": 1,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 0,
      "y": 2,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 0,
      "y": 3,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 0,
      "y": 4,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 0,
      "y": 5,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 0,
      "y": 6,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 0,
      "y": 7,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 0,
      "y": 8,
    },
  ],
  Array [
    Object {
      "filled": false,
      "value": 0,
      "x": 1,
      "y": 0,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 1,
      "y": 1,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 1,
      "y": 2,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 1,
      "y": 3,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 1,
      "y": 4,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 1,
      "y": 5,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 1,
      "y": 6,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 1,
      "y": 7,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 1,
      "y": 8,
    },
  ],
  Array [
    Object {
      "filled": false,
      "value": 0,
      "x": 2,
      "y": 0,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 2,
      "y": 1,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 2,
      "y": 2,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 2,
      "y": 3,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 2,
      "y": 4,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 2,
      "y": 5,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 2,
      "y": 6,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 2,
      "y": 7,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 2,
      "y": 8,
    },
  ],
  Array [
    Object {
      "filled": false,
      "value": 0,
      "x": 3,
      "y": 0,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 3,
      "y": 1,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 3,
      "y": 2,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 3,
      "y": 3,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 3,
      "y": 4,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 3,
      "y": 5,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 3,
      "y": 6,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 3,
      "y": 7,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 3,
      "y": 8,
    },
  ],
  Array [
    Object {
      "filled": false,
      "value": 0,
      "x": 4,
      "y": 0,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 4,
      "y": 1,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 4,
      "y": 2,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 4,
      "y": 3,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 4,
      "y": 4,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 4,
      "y": 5,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 4,
      "y": 6,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 4,
      "y": 7,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 4,
      "y": 8,
    },
  ],
  Array [
    Object {
      "filled": false,
      "value": 0,
      "x": 5,
      "y": 0,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 5,
      "y": 1,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 5,
      "y": 2,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 5,
      "y": 3,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 5,
      "y": 4,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 5,
      "y": 5,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 5,
      "y": 6,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 5,
      "y": 7,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 5,
      "y": 8,
    },
  ],
  Array [
    Object {
      "filled": false,
      "value": 0,
      "x": 6,
      "y": 0,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 6,
      "y": 1,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 6,
      "y": 2,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 6,
      "y": 3,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 6,
      "y": 4,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 6,
      "y": 5,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 6,
      "y": 6,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 6,
      "y": 7,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 6,
      "y": 8,
    },
  ],
  Array [
    Object {
      "filled": false,
      "value": 0,
      "x": 7,
      "y": 0,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 7,
      "y": 1,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 7,
      "y": 2,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 7,
      "y": 3,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 7,
      "y": 4,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 7,
      "y": 5,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 7,
      "y": 6,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 7,
      "y": 7,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 7,
      "y": 8,
    },
  ],
  Array [
    Object {
      "filled": false,
      "value": 0,
      "x": 8,
      "y": 0,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 8,
      "y": 1,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 8,
      "y": 2,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 8,
      "y": 3,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 8,
      "y": 4,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 8,
      "y": 5,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 8,
      "y": 6,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 8,
      "y": 7,
    },
    Object {
      "filled": false,
      "value": 0,
      "x": 8,
      "y": 8,
    },
  ],
]
`;
